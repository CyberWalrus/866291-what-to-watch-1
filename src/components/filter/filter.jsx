import React from "react";
import {connect} from "react-redux";
import {ActionCreator} from "../../store/filter/filter.js";
import {getGeners} from "../../store/data/selectors.js";
import {getGenreSelected} from "../../store/filter/selectors.js";
import PropTypes from "prop-types";

class Filter extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    return (
      <ul className="catalog__genres-list">
        {this.props.genres &&
          this.props.genres.map((item, i) => (
            <li
              key={i}
              className={
                this.props.genreSelected === item
                  ? `catalog__genres-item catalog__genres-item--active`
                  : `catalog__genres-item`
              }
            >
              <a
                className="catalog__genres-link"
                onClick={() => this.props.changeFilter(item)}
              >
                {item}
              </a>
            </li>
          ))}
      </ul>
    );
  }
}

Filter.propTypes = {
  genreSelected: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  changeFilter: PropTypes.func.isRequired
};
const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    genres: getGeners(state),
    genreSelected: getGenreSelected(state)
  });
const mapDispatchToProps = (dispatch) => ({
  changeFilter: (value) => dispatch(ActionCreator.changeGenre(value))
});

export {Filter};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Filter);
