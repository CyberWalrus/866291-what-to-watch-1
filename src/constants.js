const GENRE_DEFOULT = `All genres`;
const NUMBER_FILM = 20;
const NUMBER_FILM_GENRE = 4;
const SERVER_URL = `https://es31-server.appspot.com/wtw`;
const SERVER_URL_USER = `https://es31-server.appspot.com/`;
const LINK_STYLE = `inherit`;
const MAIN_FILM_ID = 1;
const PADDING_VIDEO = 25;
const REVIEW_MESSAGE = `OK`;

const ratingRadioValues = [`1`, `2`, `3`, `4`, `5`];

const RatingName = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`
};
const RatingInterval = {
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  VERY_GOOD: 10,
  AWESOME: 10
};
const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
const BodyOverflow = {
  VISIBLE: `visible`,
  HIDDEN: `hidden`
};
const FavoriteStatus = {
  ADD: 1,
  REMOVE: 0
};
const TextLength = {
  MIN: 50,
  MAX: 400
};
const RatingValue = {
  MIN: 1,
  MAX: 5
};

const DisabledStyle = {
  opacity: `0.4`,
  border: `inherit`,
  cursor: `not-allowed`
};
const OptionsVideoMin = {
  width: `280`,
  height: `175`,
  isMuted: true,
  isLoop: true,
  isControls: false
};
const OptionsVideoFull = {
  width: `100%`,
  height: `100%`,
  isMuted: false,
  isLoop: false,
  isControls: false
};

export {
  GENRE_DEFOULT,
  NUMBER_FILM,
  SERVER_URL,
  SERVER_URL_USER,
  NUMBER_FILM_GENRE,
  LINK_STYLE,
  PADDING_VIDEO,
  REVIEW_MESSAGE,
  MAIN_FILM_ID,
  ratingRadioValues,
  RatingName,
  RatingInterval,
  FilmRoute,
  OptionsVideoMin,
  OptionsVideoFull,
  BodyOverflow,
  FavoriteStatus,
  DisabledStyle,
  TextLength,
  RatingValue
};
