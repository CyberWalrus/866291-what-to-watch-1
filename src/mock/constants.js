const FilmRoute = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};
const GENRE_DEFOULT = `All genres`;
const NUMBER_FILM = 20;

export {FilmRoute, GENRE_DEFOULT, NUMBER_FILM};

