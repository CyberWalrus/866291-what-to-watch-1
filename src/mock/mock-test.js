const FILMS = [
  {
    bgColor: `#977461`,
    description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
    director: `Martin Scorsese`,
    genre: `Thriller`,
    id: 1,
    isFavorite: false,
    pageUrl: `Shutter Island`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: `8,1`,
    ratingLevel: `Very good`,
    released: 2010,
    runTime: 138,
    scoresCount: 1002557,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
    starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
    title: `Shutter Island`
  },
  {
    bgColor: `#F1E9CE`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    pageUrl: `Macbeth`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: `6,6`,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    starrings: [(`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`)],
    title: `Macbeth`
  }
];
const FILM = {
  bgColor: `#977461`,
  description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
  director: `Martin Scorsese`,
  genre: `Thriller`,
  id: 1,
  isFavorite: false,
  pageUrl: `Shutter Island`,
  srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: `8,1`,
  ratingLevel: `Very good`,
  released: 2010,
  runTime: 138,
  scoresCount: 1002557,
  srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
  srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
  srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
  starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
  title: `Shutter Island`
};
const REVIEWS = [
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  },
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  }
];

const GENRES = [
  `All genres`,
  `Thriller`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Action`,
  `Fantasy`
];
export {FILMS, FILM, GENRES, REVIEWS};
