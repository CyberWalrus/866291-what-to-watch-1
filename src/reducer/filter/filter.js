
const intialState = {
  genreFilter: `all`
};

const ActionCreator = {
  changeGenre: (value) => {
    if (
      [
        `comedy`,
        `crime`,
        `documentary`,
        `drama`,
        `horror`,
        `family`,
        `romance`,
        `sci-fi`,
        `thriller`
      ].find((item) => item === value)
    ) {
      return {
        type: `CHANGE_GENRE`,
        genreNew: value
      };
    }
    return {
      type: `RESET`
    };
  }
};

const reducer = (state = intialState, action) => {
  switch (action.type) {
    case `CHANGE_GENRE`:
      return Object.assign({}, state, {
        genreFilter: action.genreNew
      });
    case `RESET`:
      return Object.assign({}, intialState);
  }
  return state;
};

export {ActionCreator, reducer};
