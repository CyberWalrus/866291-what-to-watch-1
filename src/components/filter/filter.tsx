import * as React from "react";
import {connect} from "react-redux";
import {ActionCreator} from "../../store/filter/filter";
import {getGeners} from "../../store/data/selectors";
import {getGenreSelected} from "../../store/filter/selectors";
import {StateApp, ThunkDispatch} from "../../type/reducer";

interface PropsState {
  genreSelected: string;
  genres: string[];
}
interface PropsDispatch{
  onChangeFilter: (value: string) => void;
}
type Props = PropsState & PropsDispatch;

const Filter = ({genres, genreSelected, onChangeFilter}: Props) => {
  return (
    <ul className="catalog__genres-list">
      {genres &&
        genres.map((item, i) => (
          <li
            key={i}
            className={
              genreSelected === item
                ? `catalog__genres-item catalog__genres-item--active`
                : `catalog__genres-item`
            }
          >
            <a
              className="catalog__genres-link"
              onClick={() => onChangeFilter(item)}
            >
              {item}
            </a>
          </li>
        ))}
    </ul>
  );
};

const mapStateToProps = (state: StateApp, ownProps: Props): Props =>
  Object.assign({}, ownProps, {
    genres: getGeners(state),
    genreSelected: getGenreSelected(state)
  });
const mapDispatchToProps = (dispatch: ThunkDispatch): PropsDispatch => ({
  onChangeFilter: (value) => dispatch(ActionCreator.changeGenre(value))
});

export {Filter};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Filter);
