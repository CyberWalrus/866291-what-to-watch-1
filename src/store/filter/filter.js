import {GENRE_DEFOULT} from "../../mock/constants.js";

const intialState = {
  genreSelected: GENRE_DEFOULT
};

const ActionType = {
  CHANGE_GENRE: `CHANGE_GENRE`,
  RESET: `RESET`
};

const ActionCreator = {
  changeGenre: (value) => {
    return {
      type: ActionType.CHANGE_GENRE,
      payload: value
    };
  }
};

const reducer = (state = intialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return Object.assign({}, state, {
        genreSelected: action.payload
      });
    case ActionType.RESET:
      return Object.assign({}, intialState);
  }
  return state;
};

export {ActionCreator, reducer};
