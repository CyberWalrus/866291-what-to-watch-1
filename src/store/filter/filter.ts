import {GENRE_DEFOULT} from "../../constants";
import {Action as ReduxAction} from "redux";

enum ActionType {
  CHANGE_GENRE = "CHANGE_GENRE",
  SET_PLAY_FILM = "SET_PLAY_FILM"
}
export interface State {
  genreSelected: string;
  playFilmId: number;
}
interface ChangeGenre extends ReduxAction {
  type: ActionType;
  payload: string;
}
interface SetPlayFilmId extends ReduxAction {
  type: ActionType;
  payload: number;
}
interface ResetPlayFilmId extends ReduxAction {
  type: ActionType;
  payload: number;
}
export type Action = ChangeGenre | SetPlayFilmId | ResetPlayFilmId;

const initialState: State = {
  genreSelected: GENRE_DEFOULT,
  playFilmId: 0
};

const ActionCreator = {
  changeGenre: (value: string): ChangeGenre => {
    return {
      type: ActionType.CHANGE_GENRE,
      payload: value
    };
  },
  setPlayFilmId: (filmId: number): SetPlayFilmId => {
    return {
      type: ActionType.SET_PLAY_FILM,
      payload: filmId
    };
  },
  resetPlayFilmId: (): ResetPlayFilmId => {
    return {
      type: ActionType.SET_PLAY_FILM,
      payload: initialState.playFilmId
    };
  }
};

const reducer = (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return Object.assign({}, state, {
        genreSelected: action.payload
      });

    case ActionType.SET_PLAY_FILM:
      return Object.assign({}, state, {
        playFilmId: action.payload
      });
  }
  return state;
};

export {initialState, ActionCreator, ActionType, reducer};
