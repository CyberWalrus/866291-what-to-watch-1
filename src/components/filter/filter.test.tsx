import * as React from "react";
import renderer from "react-test-renderer";
import {Filter} from "./filter";
import {GENRE_DEFOULT} from "../../constants";
import {GENRES} from "../../mock/mock-test";

it(`Filter correctly renders after relaunch`, () => {
  const handleClick = jest.fn();
  const tree = renderer
    .create(
      <Filter
        genres={GENRES}
        genreSelected={GENRE_DEFOULT}
        onChangeFilter={handleClick}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
