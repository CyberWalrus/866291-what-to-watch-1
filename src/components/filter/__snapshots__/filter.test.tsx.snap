// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter correctly renders after relaunch 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item catalog__genres-item--active"
    key="0"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      All genres
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="1"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Thriller
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="2"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Drama
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="3"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Adventure
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="4"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Comedy
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="5"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Crime
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="6"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Action
    </a>
  </li>
  <li
    className="catalog__genres-item"
    key="7"
  >
    <a
      className="catalog__genres-link"
      onClick={[Function]}
    >
      Fantasy
    </a>
  </li>
</ul>
`;
